### ISSUE TITLE
Is there any harm in marking all my dependencies and devDependencies from my package.json set to "external"?

### ISSUE THREAD

**Issue by GitMurf** [evanw/esbuild#2297]
I have seen a few examples that use the concept of excluding library code from node modules during the compile/bundle process with the `external` option adding all package dependencies from the `package.json`. I am wondering if there is any harm in this or anything I need to be careful with? Below is the code I am using in my esbuild config script for an `electron` app. I build a list of `externalPackages` to exclude by reading the `package.json` and grabbing all dependencies and then adding to that list all the built in node modules.

Thoughts? Anything I could improve here or need to be mindful of? Bad idea? Good idea?

```js
import { build } from 'esbuild';
import { readFileSync } from 'fs';
import nodeModule from 'node:module';

let { dependencies, devDependencies, peerDependencies } = JSON.parse(readFileSync('./package.json', 'utf8'));
let externalPackages = Object.keys(!dependencies ? {} : dependencies)
    .concat(Object.keys(!devDependencies ? {} : devDependencies))
    .concat(Object.keys(!peerDependencies ? {} : peerDependencies))
    .concat(...nodeModule.builtinModules);

build({
    ...
    external: externalPackages,
    ...
}).catch(() => process.exit(1));
```

**Comment by GitMurf** [comment]
Wondering if anyone has any opinions on this or any new recommended best practices for using esbuild with Electron? Thanks!

**Comment by evanw** [comment]
Yes this is very reasonable, and is already recommended when bundling for node (which Electron uses): https://esbuild.github.io/getting-started/#bundling-for-node. There's even a dedicated option for this now: https://esbuild.github.io/api/#packages. I don't have any additional Electron-specific recommended best practices.

**Comment by hyrious** [comment]
There're 3 different entries that electron may use:

* The Node.js part can be safely bundled with `--packages=external` since it can access all dependencies at runtime from node_modules.

* The [preload.js](https://www.electronjs.org/docs/latest/tutorial/tutorial-preload) runs in the browser's script environment with a special top-level `require` function. You can assume it is loaded via `<script src="preload.js">`. In which case you should bundle it in IIFE format (which is the default in esbuild when bundling) with `--external:electron`.

* The other browser scripts work the same as in a real browser. You can choose any convenient way to bundle and load it. For example I may prefer load a native ESM file with splitting enabled for better performance with `--format=esm --splitting` then load it in `<script type="module" src="bundle.js">`.
